### PR TITLE
New function to log object graph

### PR DESCRIPTION
**Pull request by beta-ziliani**
Produces beautiful graphs like:

```mermaid
graph LR
  0x101832e80["0x101832e80 C"]
  0x101832e70["0x101832e70 C"]
  0x10182be80["0x10182be80 A"] --@c,0--> 0x101832e80
  0x10182be40["0x10182be40 B"] --@c,0--> 0x101832e80
  0x10182be20["0x10182be20 (class 0)"] --@(field 0),1--> 0x10182be40
  0x10182be60["0x10182be60 Array(B)"] --@(field 16),2--> 0x10182be20
  0x10182be80["0x10182be80 A"] --@bs,3--> 0x10182be60
  0x10182be00["0x10182be00 B"] --@c,0--> 0x101832e70
  0x10182be20["0x10182be20 (class 0)"] --@(field 8),1--> 0x10182be00
```

### CODE REVIEW

File: src/perf_tools/mem_prof.cr [hunk 1]
```diff
@@ -326,6 +344,144 @@ module PerfTools::MemProf
...
+    GC.collect
+    stopping do
+      type_id = T.crystal_instance_type_id
+      alloc_infos = self.alloc_infos
+      pointers = alloc_infos
+        .select { |_, info| info.type_id == type_id }
+        .map { |ptr, _| {ptr, 0} }
+
+      pointers = REF_LIMIT == 0 ? pointers.to_h : pointers.first(REF_LIMIT).to_h
+      original_pointers = pointers.dup.map { |ptr, _| ptr }
```
Comment (ysbaddaden) [inline src/perf_tools/mem_prof.cr:407]: Functional-like programming is very nice, but we might be manipulating lots of data in this method —I guess? Maybe we can inline the manipulation and avoid intermediary representations?

```crystal
pointers = {} of Pointer => Int32

alloc_infos.each_value do |info|
  next unless info.type_id == type_id
  pointers[ptr] = 0
  break if REF_LIMIT > 0 && pointers.size >= REF_LIMIT
end
```

Also, no need to `dup` + nice method:

```crystal
original_pointers = pointers.keys
```

**Review by ysbaddaden — APPROVED** [review]
I left a suggestion to reduce avoid some allocations (if manipulating lots of data), and maybe we can optimize a couple more things later.

Another suggestion would be to split building the stack in one method, from generating the mermaid graph in another (maybe).

**Comment by beta-ziliani** [comment]
> Another suggestion would be to split building the stack in one method, from generating the mermaid graph in another (maybe).

Yes, this is an issue with the tool, see #21 .

**Comment by beta-ziliani** [comment]
> Addressing comments from *core* review

*core review:* it is said of a code review from a core team member 🤭

**Review by ysbaddaden — APPROVED** [review]

**Pull request merged**